accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/authorizableinstaller/impl/PrefetchingUserManager.java

```diff
@@ -0,0 +1,196 @@
+package biz.netcentric.cq.tools.actool.authorizableinstaller.impl;
+
+import static biz.netcentric.cq.tools.actool.history.PersistableInstallationLogger.msHumanReadable;
+
+import java.security.Principal;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+import javax.jcr.RepositoryException;
+import javax.jcr.ValueFactory;
+
+import org.apache.jackrabbit.JcrConstants;
+import org.apache.jackrabbit.api.security.user.Authorizable;
+import org.apache.jackrabbit.api.security.user.Group;
+import org.apache.jackrabbit.api.security.user.Query;
+import org.apache.jackrabbit.api.security.user.QueryBuilder;
+import org.apache.jackrabbit.api.security.user.User;
+import org.apache.jackrabbit.api.security.user.UserManager;
+import org.apache.jackrabbit.oak.spi.security.user.UserConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import biz.netcentric.cq.tools.actool.helper.Constants;
+import biz.netcentric.cq.tools.actool.history.InstallationLogger;
+
+/**
+ * <p>
+ * Special user manager that prefetches
+ * <ul>
+ * <li>all groups, system users and anonymous (but not regular users)</li>
+ * <li>all memberships between the preloaded authorizables</li>
+ * <ul>
+ * upon creation. 
+ * </p>
+ * <p>
+ * The membership relationships are cached in lookup maps to quickly be able to query the repository state for both
+ * authorizable.declaredMemberOf() (this we call in this class also) and group.getDeclaredMembers(). Having called declaredMemberOf() for
+ * all groups of the repository has also retrieved all memberships (expect regular user memberships), hence the method getDeclaredMembers()
+ * does not have to be called anymore. This is particularly useful because that way the AC tool does not have to iterate over a potentially
+ * huge number of users in production to then only filter out a few relevant groups in the end (the AC Tool does not touch user memberships
+ * to groups).
+ * </p>
+ */
+class PrefetchingUserManager implements UserManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PrefetchingUserManager.class);
+
+    private final UserManager delegate;
+    private final Map<String, Authorizable> authorizableCache = new HashMap<>();
+
+    private final Map<String, Set<String>> nonRegularUserMembersByAuthorizableId = new HashMap<>();
+    private final Map<String, Set<String>> isMemberOfByAuthorizableId = new HashMap<>();
+
+    public PrefetchingUserManager(UserManager delegate, final ValueFactory valueFactory, InstallationLogger installLog)
+            throws RepositoryException {
+        this.delegate = delegate;
+
+        long startPrefetch = System.currentTimeMillis();
+        Iterator<Authorizable> allGroupsAndSystemUsersAndAnonymous = delegate.findAuthorizables(new Query() {
```

Comment:
`authorizablesToPrefetch` instead of `allGroupsAndSystemUsersAndAnonymous`?

Reply:
done